joeynmt/prediction.py

```diff
@@ -117,15 +122,23 @@ def validate_on_data(model: Model, data: Dataset,
             # run as during inference to produce translations
             output, attention_scores = run_batch(
                 model=model, batch=batch, beam_size=beam_size,
-                beam_alpha=beam_alpha, max_output_length=max_output_length)
+                beam_alpha=beam_alpha, max_output_length=max_output_length,
+                n_best=n_best)
 
             # sort outputs back to original order
             all_outputs.extend(output[sort_reverse_index])
             valid_attention_scores.extend(
                 attention_scores[sort_reverse_index]
                 if attention_scores is not None else [])
 
-        assert len(all_outputs) == len(data)
```

Comment:
Replace with assertion of new length. Have you considered making the outputs a structured object rather than a flat list? As a flat list, it is not obvious which outputs correspond to which references (in the evaluation below), so they should rather be treated separately (i.e. separate evaluation for translations of rank0, rank1, etc, and written into separate files which all have the length of the references).